sklearn/linear_model/tests/test_ridge.py

```diff
@@ -1494,11 +1514,128 @@ def test_ridge_sag_with_X_fortran():
+
+
+@pytest.mark.parametrize("fit_intercept", [True, False])
+@pytest.mark.parametrize("alpha", [1e-3, 1e-2, 0.1, 1.0])
+def test_ridge_ground_truth_positive_test(fit_intercept, alpha):
+    # Test whether Ridge converges to the same solution
+    # with positive=True and positive=False
+    # when the ground truth coefs are all positive
+    rng = np.random.RandomState(42)
+    X = rng.randn(300, 100)
+    coef = rng.uniform(0.1, 1.0, size=X.shape[1])
+    if fit_intercept:
+        intercept = 1
+        y = X @ coef + intercept
+    else:
+        y = X @ coef
+    y += rng.normal(size=X.shape[0]) * 0.01
+
+    results = []
+    for positive in [True, False]:
+        model = Ridge(
+            alpha=alpha, positive=positive, fit_intercept=fit_intercept, tol=1e-10
+        )
+        results.append(model.fit(X, y).coef_)
+    assert_array_almost_equal(*results)
+
+
+@pytest.mark.parametrize(
+    "solver", ["svd", "cholesky", "lsqr", "sparse_cg", "sag", "saga"]
+)
+def test_ridge_positive_error_test(solver):
+    alpha = 0.1
+    X = np.array([[1, 2], [3, 4]])
+    coef = np.array([1, -1])
+    y = X @ coef
+
+    model = Ridge(alpha=alpha, positive=True, solver=solver, fit_intercept=False)
+    with pytest.raises(ValueError, match="does not support positive"):
+        model.fit(X, y)
+
+    with pytest.raises(ValueError, match="only 'lbfgs' solver can be used"):
+        _, _ = ridge_regression(
+            X, y, alpha, positive=True, solver=solver, return_intercept=False
+        )
+
+
+@pytest.mark.parametrize("alpha", [1e-3, 1e-2, 0.1, 1.0])
+def test_positive_ridge_loss(alpha):
+    # Check ridge loss consistency when positive argument is enabled
+    X, y = make_regression(n_samples=300, n_features=300, random_state=42)
+    alpha = 0.10
+    n_checks = 100
+
+    def ridge_loss(model, random_state=None, noise_scale=1e-8):
+        intercept = model.intercept_
+        if random_state is not None:
+            rng = np.random.RandomState(random_state)
+            coef = model.coef_ + rng.uniform(0, noise_scale, size=model.coef_.shape)
+        else:
+            coef = model.coef_
+
+        return 0.5 * np.sum((y - X @ coef - intercept) ** 2) + 0.5 * alpha * np.sum(
+            coef ** 2
+        )
+
+    model = Ridge(alpha=alpha).fit(X, y)
+    model_positive = Ridge(alpha=alpha, positive=True).fit(X, y)
+
+    # Check 1:
+    #   Loss for solution found by Ridge(positive=False)
+    #   is lower than that for solution found by Ridge(positive=True)
+    loss = ridge_loss(model)
+    loss_positive = ridge_loss(model_positive)
+    assert loss <= loss_positive
+
+    # Check 2:
+    #   Loss for solution found by Ridge(positive=True)
+    #   is lower than that for small random positive perturbation
+    #   of the positive solution.
+    for random_state in range(n_checks):
+        loss_perturbed = ridge_loss(model_positive, random_state=random_state)
+        assert loss_positive <= loss_perturbed
```

Comment:
As the optimisation problem is convex, we could check first order conditions. I find this check against a random point rather weak. @agramfort What do you think?

Reply:
@lorentzenchr  I implemented logic to check whether every component of Jacobian in positive solution are no less than zero, but some component has small minus value even when using small tolerance in L-BFGS-B solver. If I fixed the logic to check whether Jacobian is greater than small minus value (`-1e-4`), the test passed. To check loss function against a random point is weak but proper idea in this test. What do you think about it?

Reply:
@lorentzenchr Fixed logic to check every component of Jacobian of loss for solution found by `Ridge(positive=True)` is no less than `-1e-4` (zero minus small value).

Reply:
The 1. order and 2. order conditions with constraints/bounds are a bit more complicated to check. Some of the coefficients may be on the boundary. Maybe we can skip this test altogether and rely on lbfgs? I mean we do not need to check the functionality of lbfgs in scikit-learn again.

Reply:
@lorentzenchr 

> The 1. order and 2. order conditions with constraints/bounds are a bit more complicated to check.

I agree. I think it is proper to check against random points under positivity constraints. Although this test is weak, it should be allowed in this situation. So I rolled back it to the original logic. Would you check it once again?